### ISSUE TITLE
Gnu PGP support?

### ISSUE THREAD

**Issue by jonasbb** [nylas/N1#19]
I just stumbled on your new e-mail client. It looks really nice. I would like to try this app but I need PGP support. I could not find anything in the source code, so I guess it is not supported.

Any plans to add this feature to your client or sync platform?

**Comment by dmvaldman** [comment]
:+1: 

I'm guessing this will take some effort. But this is a great idea for a plugin!

**Comment by gtramontina** [comment]
I'd love to contribute with such plugin. Using https://github.com/openpgpjs/openpgpjs should be quite straightforward I believe.

**Comment by dmvaldman** [comment]
Or just straight up GnuPG which is what GPG tools (https://gpgtools.org/) uses for their Apple Mail extension. Since Nylas is a desktop app, you don't need to work within the confines of a browser.

**Comment by teamcoltra** [comment]
:+1:

**Comment by frdmn** [comment]
:+1:

**Comment by gtramontina** [comment]
I agree with using a battle-tested implementation is the way to go... My point on suggesting `openpgpjs` was more related to cross-platform.

**Comment by bengotow** [comment]
Hey folks—this is a great idea. It's definitely the kind of thing we'd like to see implemented as a plugin. I'm not sure how the GPG extension for Mac Mail works, but something similar should work in N1 because we have access to the filesystem, can run shell commands, etc. I've added this to the plugin ideas section of our public Trello board:  https://trello.com/b/hxsqB6vx/n1-open-source-roadmap. Feel free to hop on our community Slack channel if you'd like to try building the plugin!

**Comment by whitj00** [comment]
+1

**Comment by tracker1** [comment]
Just another idea along the same vein would be to add keybase.io support for public/private key for the user, and public key lookup for contacts.

**Comment by alexanderadam** [comment]
@whitj00 I am not part of nylas but I think that the +1 won't affect the plan here. That's why the issue was closed. So you would rather have to make your +1 on the referenced trello link.

@tracker1 I would say getting keys from keyservers is another issue and doesn't have to be part of the initial implementation. Especially because some native GPG frontends have support for keyservers anyway.

**Comment by xaxes** [comment]
+1

**Comment by dcposch** [comment]
**I'm trying to PGP support to Nylas.**

I'll do as much as I can within the module system. Beyond that, though, I may have to modify it my fork of N1 in deeper ways.
There are a few challenges that make encryption hard to just layer on as just a plugin:

* You don't want to save drafts unencrypted
* You don't want to upload attachments unencrypted
* Doing it right requires integration in a lot of places. A row in the inbox should a lock icon if it's an encrypted mail. The Send button should show a lock icon when all recipients have keys. The Compose screen needs to help the user find keys for recipients, looking them up in Keybase. A lot of screens require hooks to transparently decrypt and verify mail.

There are additional challenges that make it hard for N1 in particular:
* N1 currently delegates search to the Sync Engine API. For a PGP client, search has to be on the client.
* N1 assumes that the Sync Engine has access to plaintext in a few other places, for example for the inbox preview snippets.

**Comment by mbilker** [comment]
I am working on this. @dcposch thank you for the goals I should be working on.

**Comment by mbilker** [comment]
My package is live https://github.com/mbilker/email-pgp

**Comment by teamcoltra** [comment]
@mbilker Is this a fork?

**Comment by mbilker** [comment]
@teamcoltra no, it is a N1 plugin.

**Comment by sicofante** [comment]
Whatever the solution, encrypting should be seriously considered from the UI point of view (take a look at Mailpile). It should be part of the base design. The user should have a way to know which of his contacts can decrypt messages and the system should automatically encrypt every message sent to them without asking. For instance.

**Comment by mbilker** [comment]
@sicofante I could add a option to encrypt every outgoing message, and how to encrypt the message. Whether its straight encrypt where you need the private key to read the email or a signed message to verify the message comes from the owner.

I am working on decoding the Keybase sig chain to verify the user's tracked users. Because the tracked users are based upon an encrypted message using the public key, I can verify the signature matches the locally stored public key from the user with `gpg --verify <message file name>`. Figuring out how to do that in openpgp.js is hard.

**Comment by mbilker** [comment]
@sicofante Here is my preferences tab as it stands:

![2015-12-02-111226_1280x800_scrot](https://cloud.githubusercontent.com/assets/1266541/11536262/9ca34fce-98e5-11e5-96a2-84b0b2a6cfc3.png)

**Comment by sicofante** [comment]
I'm afraid I don't quite understand the technical explanations you have given, but if you can achieve what I propose, it would encourage the use of PGP, which should be a primary goal of Nylas N1. Thanks for your work.

**Comment by mbilker** [comment]
Currently the encryption aspect is implemented, but I need to implement decryption with proper storage for private/secret keys. I also need to find a way to handle quoted previous emails.

**Comment by Loscillo** [comment]
Hi,

First of all thank you for your efforts to integrate this functionality. However, I used Enigmail and I did not feel that he was storing the key (just the password for a defined period). It is not possible to do something similar here?

**Comment by sicofante** [comment]
I haven't used Enigmail myself, but all I've read is it's not trivial to use. A solution that's not straightforward (think Signal for Android or Mailpile for e-mail) and works only for geeks is not good.

**Comment by Loscillo** [comment]
I have not found that Enigmail was difficult to use, but if you have read that it was the opposite I understand.

**Comment by sicofante** [comment]
It won't be difficult for you or for me, but for ordinary users, the process should be transparent. If I send an email to someone who uses PGP, that message should be automatically and transparently encrypted. If not, the message should not be encrypted. I should never see any non-decrypted text. No gibberish in sight. Ever. Zero configuration, or almost. That's what Signal offers for encrypted communications in Android.

Make it easy to tell secure messages from ordinary ones, make it easy to invite your friends, family and colleagues to use this new e-mail privacy-minded service and you'll have a hit.

Ask users to go through steps to encrypt a message and [almost] nobody will use it. Exactly what happens today with Enigmail.

**Comment by mbilker** [comment]
Currently the steps to encrypt an email with my plugin: Create email -> Click Encrypt -> Type User's Keybase name -> Click Submit -> Message is Encrypted shortly thereafter -> Send Email

**Comment by mbilker** [comment]
My task right now is to decode the sig chain to pull out tracking users to eliminate typing a user's keybase name.

I am not an algorithms person, so I will need to find some algorithm that takes the tracked users and the most recently used usernames to compile a list to make the selection process quicker. I may even allow binding of emails to usernames, since keybase doesn't provide a way to search by email to user (for obvious privacy reasons).

**Comment by mbilker** [comment]
Hmm. I will continue to use #19 (this issue) as the primary issue to address any concerns regarding PGP in N1.

**Comment by mbilker** [comment]
Could someone who uses any kind of PGP send a PGP encrypted email to me@mbilker.us with the public key (fingerprint: 69AD F8AE B6C8 B5D1):
```
-----BEGIN PGP PUBLIC KEY BLOCK-----
Version: GnuPG v2

mQINBFVtOYcBEADg2AAPtuo47Xkx/uY44xHVd6LjawJ8bIPNGQGb3R1mzqutlA5P
GfCpS3VXfiwSwzJv2OGZFkyMgK+GcX/KbsOTKkIkc9A1d9h6QGqBNTinvsmo7nvG
uYX0jNTwMSTALW0GR2DGgWRpSDU6o5r2XnrQEGXzjLSGW/rB/Ati4ZCg1vDd6m1t
XLhUbh4kPxkp50M0UIyZxddgsDKwqQwyCy3TENejJ+edT/jXBBgBZ1TNxEGcm5mG
UrdUrgomiezf8l8Z3Uwaq9cJnbLiYQ/L4ZTVoQdTXSLXWbhMC8a82RqMQOExRKxM
T6PCy4sLmd4OfbKyVnKZ+0dWfd6mwG6e9GXsPpNKG48B/P6GxzfqAP7/Vhsj47MT
N1zw0+e9L1UygqHNdO7SK6LIVUZhDXqcxgmX6wjiRmgYk8JgPvgMJFVg9eih1R/T
2owTEp8xs9AHlQ/XiGjt8OtaWRxVH51rM0jDH25iOm5TbQXnwSzfJbp3E3ImAqmG
rXuGRCRwtPlbw6mz+NzKfknrUTJQQ56S+G9JX7sgszucuUC5xlteac2UczzO1a1e
O2Fioo3AmGXw2jpKdHt4UcequJtoDiCkR/bqkojOf6fgSta4kgz4vUywa188WlAG
Tkt6p7bigYxJE4P7CQTpVdMsBBux0AxJWLCuxPvCM9ASFoy/3Bsn0dK4sQARAQAB
tBtNYXR0IEJpbGtlciA8bWVAbWJpbGtlci51cz6JAj0EEwEKACcFAlVtOYcCGwMF
CQPCZwAFCwkIBwMFFQoJCAsFFgIDAQACHgECF4AACgkQaa34rrbItdE+rw//cx0P
5Fa3wrz0AD8GfUwyz9ibZL4lXriiFbGiG0TCgak+f8Vj4ZFEWrKy3RjzVkFMSdYE
xQlcs4xBWq+D8ldOS5ZwG5C6GP8VfWX56e8HG4zLGJOsQrSmhoAAvtKo67BoDvi/
fOO4a1hooYLgh4zafQCbBedYx+ylILhlJzVXH/krGMkd+GgiY0lxUnVwuIU7RWI9
r/Qe3LygjBlarR5S6W10C/kK1k8OFrxfMUbypc7LjCLT6rj9CLfQBOEtdcmL43yq
kjN2IUkl8AsJEaVaKVoTTO/xMnb8aO+8IUuYZjgnw4dr7kdOPh8svlHlU9iDzIFs
ict3S/SjS+v9RZHBtgf1StWiuZDbehpgOxdw4MhzN4BtOaOEHJ6vMdmDFZwnDLrR
q+QdHxUyjyMcCGrhgmGboQ2SOCfWRxLbhFRZPIZ72Lu2l817nNkxJbL3hl+nQXyZ
ZE40j1EyLaRTBK326QEhmbnq6R7jmiEOgu+H/4m6wgzQL/iIwcX2SVkDX6lwvg4n
f5WLvzinLjVwp9lMofa1ciZj/jH1UryXG6qZnrdz4SjVM6tZD3d+vfwyQrruq09g
8d0ldP/ccMKFCf64NYnhTCKcqpHZ1/nPN1Ta58jHA8XtRB103Iynm3CQ21xiT8OZ
9qKRipiSIcA8CNt8VkDgQTSqlaHyIr25nBH36aS0J2tleWJhc2UuaW8vbWJpbGtl
ciA8bWJpbGtlckBrZXliYXNlLmlvPokCNAQTAQoAHgUCVW05hwIbAwMLCQcDFQoI
Ah4BAheAAxYCAQIZAQAKCRBprfiutsi10WSeEADEP0Zq+QRVRfz2FeQ7zL5j9tcD
R4kjcowXvuR2Hdjx8I+g+tvBzcHOuRDikHyfQR093yOlbx/2aqLLxLwyKmQu+PSP
swKHvBVQEnV6Gskjr6/w5VBtt7OmPg3yeWLgixn1etQ1g71KOG1J0wOULmqfnZT/
LNAsy9nA82I2mTiIlaRE7edlO9NV4JuynrwHYFpK6o9EoBN4rTJZtm8lF/8FuIoo
RGHCCbFa+UqdLLbfCc0J3zsmQ6n4z3DDWwxlnon/3821cufnKOrJ/ueKRY2qv5aT
/OPDlOS7n6E/3Gx6wNNjatNeNbmNRRa/QGoxwiSui3wIzOaaADHs7ptyZ8E50N4g
L52OcnGun/ehxmmYorZXjzmiJkRzS/iNElpEXC3Gbcf1A+xb/wRbwRPDclYmtTbw
c6eOpARMmmFZzEzJb3UIn5TfZEokBwQ2igcZyDFFqZvi/OKz37/Wyn/IODQFYZrY
1FeIPtMB7NIjnY1Vgi2jt2ihZpVf96J9j1ebe4RYueQfs+Rx9F7C5cbsAQky5zOk
J7hXnGzqc5q5LJz0/5PynTJxJkObCNKGkKrAsaFGVbdzUrzoDH3PljTb8kjS1FDx
rQ9h58uolG55ddMOFdLP5pga1zNLktin3Atu81zsVd3JzN0RLrZTirGO7sPSFUv0
5WJfrqjbUkef9GFUEbkCDQRVbTmHARAAu/HH86/awppFoWfP7IBRd1ORRUdn7rZq
0ZoxmEpLiU6+WC2mRRC8Qs5C/NrHviEnZCNZLiy9AOXSNq5MwCr+QwVZ57A/yqHF
Hsnccqj5DolOACo7Zr6wjaA1m1SCEucKSunidB1FIJFfuUv/K07cPkvF8KvjX57+
JsBgLxEAC5ZYW1qy5lq7oZJlaeOxYw7kGJgs5vJg8uKmpjFz2l+00/ssWscCSuox
tvKR0ac+nqCg7k8Jy8Tn6X7udbEbbes4buYa1tFyINAfR/C2oM5GF4r0FcV2uhCV
dw9N1aHsvL3xQiKcf4rRWQFRskA8lXUismrejoJkV2octgQmjXLzv2X3o2fK/RoL
MUJa19xl3SqmXF2jUzudMWE2rXXS1Sz+QiKDSsDXBrUzeGvWQxWrJKxKYYPV6VmJ
WaYIJNh0IszrNsBNGFBCHced2JPhsIHV4OfQUEbjE2Lri9iHVq1yQMjxExt7GTAe
+CWU/N/MIvN6ciDLh5OxgLT8uPTK3B1rh0oOsCWPsa3eCAlGgAipOl8TJk/MhGYU
KKkFfs4ko3VacZzkODH9gJfc5/DPDuUSbWHV83UOSgA9ucJzaOkvuEmQ1Sv2SJdP
ZytZGAebMfNWWaXQYwVZm/zI9m1SfhJNQocLbGadvRHlQ8Yob8Wj2vdSw1k+Wwmi
ZaxRtH9Eo0cAEQEAAYkCJQQYAQoADwUCVW05hwIbDAUJA8JnAAAKCRBprfiutsi1
0avzEACFLM0wev7K4fLq1/HkeaQRl2/pd2dFMrgD5dbI8O4xeQpVSJtnShnIqHOn
hkihYN3KLvZxcTaiy9gOQfXEfTFzvsptnjUO4Lnk5Blu7bW9Z6Y0ndE+H/PebDgk
7yGZIJM3QB7+7jtLkWaUP3ok2zpFVhnFODWf+gfPQ4HYqqDkRZ8HWRJ8OhKtySYN
9hRJDaoAwlKbY5G5eNqPgVWL96J1bE76GntcOCyj7/gsEJ9VqQBCjjjbNr0sp/1j
AhetuG/VavtD141Pmdb4lyQx86grGuYFNEzBwSlbVm15IVzlUFUHykJAcTcSr5nT
A45GBeoqEhD/kmpHUm/P10ONxt1qkjJ3q8q40UnPW+POBcfhTMiaM22XIxdxzOyn
Aa5h+RCZP7nzPmkMb8pQeRCq8HXq7xRzM+ZhPwoIoJl+wc4noi6kKuomoyG25aDm
Ht9LeehJQteyAdOCaWkSK1Ec3L5e9dur9vO43TYMmlzK29re1zNN5BHnKP2cHxmJ
EsrtsclMBYEwsqjGeEoBWD0mc6V1nOeddnMDAfLGACJgF4XBO5u3a64quWYl7ikK
/AN33po70UHkKabr2aoVVmZ7FLbwTlQX/tQbQToUnvKH5xpS0/1J19UQFhsow3GC
I9qMmFUMOXkQgblYpxFHW3k/CMgtfP5m2kmerT7H1oU3tCOi37kCDQRVdQWOARAA
oSXrfRHrlbUiqNM0pQY4s4A80y/XhKbLDUwMIrnF9FAq88wEZe8S0MOAt3RSco7z
6irf4QmeqiDk8z6is8mKHN9kzcO9n7oLtM7AGrvMEsHwPF7rNY4gAOg31YEIxJc/
T6rTGqizFipklXshDo3lHQOTay1+aOV91FOvEFKB8r2BboqU3kJpi29rA/dBzCgg
nW6BBFp5Iigpc6A/T+QJf7oALjUILHi4szboL+O9Zb1/TPoUDvw5F/Rl0rgFV+ld
4+8Puj7zJOHatB696uak3Po3nuGZ5L7qB0VLseL8EUa3XWAIFKqfOM1OAS1Cc05R
6FPZHoyUUIhBoVCGZ4l2PPNwGRGbqbcmjQizv8Kwcbql83BRhHI0rEp5pggU6hkz
6ulhPvh4DweQfL+JzDwEZEnqyPSEbhcTFfrFP6YJAYPkh5YSE01QAqTaBV4xUEGI
diQG3M/EEmY8HFqk7ESuTWxKVr9IZjFcCY/C25f6gcCf2Vv6qdqTl3242ZFBLcgZ
2zmKNeNDQhOGIzwKSSU7WwezcieYLaR5M/5E+RTyvN0lIVFdSHJzIAE4Oq6+ybWT
G20YBfhISYjGpskOpy2AzWuKf3Z9mSnsXpu35Cnpq2+FPq56W/rTK9xD6eu0e09S
DCtntNnteHwGeBrSSLzWZ5DeyOKvxXp6ZX7uAx/UMSUAEQEAAYkCJQQYAQoADwUC
VXUFjgIbDAUJAeEzgAAKCRBprfiutsi10YxVEACVKAQ2iv04OOCLTJmHOMj3Vy1z
s5SdJPZOpGX9wv7CcZs1AUwvAFweVly4786900qhT+/cqg2LmzgRepe/HYOgG0hD
whDIRWraqcDBaFTFbT4RiIf7X/mXZ9hbkhhRLhJRQdRerWQeio2rWJLjCY2M2LC3
FgE/OqgaG8VthsTnEMsfb1s6YNLWZWQ0DtKVQdhRchXl0mJjTUwC1b8jIm/9CFeo
XWZQUBdUVr50Ox4PmsBoLnuUJx+bKwxrCZF4ksTZHvJtYkQDoknZeE34aswSPj2S
Pe3SaVFi9vNGY90WmdWprfUbDxNA9U7LI4Qs1tU/XnFx8/BXH3nnDBC+9IZS3kAJ
dkdsmiIwHA4LMXpSYxwOVmkFboZIQQDSt6FCWwoLYlyz+1bCaW1oHR6Chgk7/2UQ
jngOmX2YKQG9yKgFpXHIS4YHlAViHJnz0CHs7s7p4ug9Aayu5hBIJPGgn4w3fc+w
YfZTiJ1ZSlRi7gDw99I+9TEf29kccZEx1/HYMOz+4L1f8WX/SsPXR5fppbUMvFjO
abN3LFoFRMNxzC59QdV5cThlOfTrcPnwnlXd+ub72EiOodIDTT8GjfhWTmT+S35Y
McXJotenrtN1dMZb/pJekX3GP/OBGDFmjGkoI+b/ppBhaS6MMZhAQ2VFDTLN7wWk
ukfHFWL0vn1fITqc1w==
=YITT
-----END PGP PUBLIC KEY BLOCK-----
```

**Comment by mbilker** [comment]
I need to get a grasp how the PGP emails appear in N1. From what I saw in the screenshots for GPGTools, there are 2 attachments. The first attachment is a dummy, but it has a MIME type set. The second attachment is the actual encrypted message, with the file extension being `.asc` dictating its a PGP encrypted message.

**Comment by sicofante** [comment]
I'll set up an account to help you. Give me a few hours.

**Comment by nickolai** [comment]
What's the state of Gnu PGP support and why is the issue closed?

**Comment by alexanderadam** [comment]
@nickolai people are opening GPG/PGP tickets all the time. But as you can see in this ticket there are [references from and to still open tickets](https://github.com/nylas/N1/issues/19#ref-issue-111963798).

So the ticket currently handling GPG/PGP is #96, although people seem to talk in parallel anyway.

**Comment by willricketts** [comment]
@nickolai The issue has been solved with https://github.com/mbilker/email-pgp

**Comment by mbilker** [comment]
@willricketts Still in development. I slowed down development because of school work.

**Comment by mbilker** [comment]
I moved this conversation to #96
